File: src/cs/Ssh.Tcp/SshServer.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Events;
using Microsoft.DevTunnels.Ssh.Messages;

namespace Microsoft.DevTunnels.Ssh.Tcp;

/// <summary>
/// Enables accepting SSH sessions on a TCP socket.
/// </summary>
/// <remarks>
/// It's possible to create an `SshServerSession` over any `Stream` instance;
/// this class is merely a convenient helper that manages creating sessions
/// over `Stream` instances obtained from a `TcpListener`.
/// </remarks>
public class SshServer : IDisposable
{
	private readonly SshSessionConfiguration config;
	private readonly CancellationTokenSource disposeCancellationSource;
	private readonly object sessionsLock = new object();
	private readonly List<SshServerSession> sessions;
	private readonly List<SshServerSession>? reconnectableSessions;
	private readonly TraceSource trace;

	public SshServer(SshSessionConfiguration config, TraceSource trace)
	{
		if (config == null) throw new ArgumentNullException(nameof(config));
		if (trace == null) throw new ArgumentNullException(nameof(trace));

		this.config = config;
		this.trace = trace;
		this.disposeCancellationSource = new CancellationTokenSource();
		this.sessions = new List<SshServerSession>();

		if (config.ProtocolExtensions.Contains(SshProtocolExtensionNames.SessionReconnect))
		{
			this.reconnectableSessions = new List<SshServerSession>();
		}
	}

	public IReadOnlyCollection<SshServerSession> Sessions => this.sessions;

	public event EventHandler<SshAuthenticatingEventArgs>? SessionAuthenticating;
	public event EventHandler<SshServerSession>? SessionOpened;
	public event EventHandler<SshRequestEventArgs<SessionRequestMessage>>? SessionRequest;
	public event EventHandler<SshChannelOpeningEventArgs>? ChannelOpening;
	public event EventHandler<SshRequestEventArgs<ChannelRequestMessage>>? ChannelRequest;
	public event EventHandler<Exception>? ExceptionRasied;

	public SshServerCredentials Credentials { get; set; } = new SshServerCredentials();

	/// <summary>
	/// Gets or sets a factory for creating TCP listeners.
	/// </summary>
	/// <remarks>
	/// Applications may override this factory to provide custom logic for selecting
	/// local port numbers to listen on for port-forwarding.
	/// </remarks>
	public ITcpListenerFactory TcpListenerFactory { get; set; }
		= new DefaultTcpListenerFactory();

	public async Task AcceptSessionsAsync(
		int localPort,
		IPAddress? localAddress = null)
	{
		localAddress ??= IPAddress.Any;
		string portPrefix = localAddress.Equals(IPAddress.Any) ?
			"port " : localAddress.ToString() + ":";

		TcpListener listener;
		try
		{
			listener = await TcpListenerFactory.CreateTcpListenerAsync(
				localAddress,
				localPort,
				canChangePort: false,
				this.trace,
				CancellationToken.None)
				.ConfigureAwait(false);
		}
		catch (SocketException sockex)
		{
			this.trace.TraceEvent(
				TraceEventType.Error,
				SshTraceEventIds.ServerListenFailed,
				$"{nameof(SshServer)} failed to listen on {portPrefix}{localPort}: {sockex.Message}");
			throw;
		}

		localPort = ((IPEndPoint)listener.LocalEndpoint).Port;

		using (this.disposeCancellationSource.Token.Register(listener.Stop))
		{
			this.trace.TraceEvent(
				TraceEventType.Information,
				SshTraceEventIds.ServerListening,
				$"{nameof(SshServer)} listening on {portPrefix}{localPort}.");
			try
			{
				while (true)
				{
					var stream = await AcceptConnectionAsync(listener).ConfigureAwait(false);
					if (stream == null)
					{
						// The server was disposed.
						break;
					}

					this.trace.TraceEvent(
						TraceEventType.Information,
						SshTraceEventIds.ServerClientConnected,
						$"{nameof(SshServer)} client connected.");

					var session = new SshServerSession(
						this.config, this.reconnectableSessions, this.trace);
					session.Credentials = Credentials;

					lock (this.sessionsLock)
					{
						this.sessions.Add(session);
					}

					session.Authenticating += (s, e) =>
					{
						SessionAuthenticating?.Invoke(s, e);
					};
					session.Request += (s, e) =>
					{
						SessionRequest?.Invoke(s, e);
					};
					session.ChannelOpening += (s, e) =>
					{
						ChannelOpening?.Invoke(this, e);
						if (e.FailureReason == SshChannelOpenFailureReason.None)
						{
							e.Channel.Request += (cs, ce) =>
							{
								ChannelRequest?.Invoke(cs, ce);
							};
						}
					};
					session.Closed += (s, e) =>
					{
						lock (this.sessionsLock)
						{
							this.sessions.Remove(session);
						}
					};
					SessionOpened?.Invoke(this, session);

					var task = Task.Run(async () =>
					{
						try
						{
							await session.ConnectAsync(stream, CancellationToken.None).ConfigureAwait(false);
						}
						catch (SshConnectionException ex)
						{
							await session.CloseAsync(ex.DisconnectReason, ex).ConfigureAwait(false);
							ExceptionRasied?.Invoke(this, ex);
						}
						catch (Exception ex)
						{
							await session.CloseAsync(SshDisconnectReason.ProtocolError, ex)
							.ConfigureAwait(false);
							ExceptionRasied?.Invoke(this, ex);
						}
					});
				}
			}
			catch (Exception ex)
			{
				ExceptionRasied?.Invoke(this, ex);
			}
			finally
			{
				listener.Stop();
			}
		}
	}

	protected virtual async Task<Stream?> AcceptConnectionAsync(TcpListener listener)
	{
		TcpClient tcpClient;
		try
		{
			tcpClient = await Task.Run(() => listener.AcceptTcpClientAsync())
				.ConfigureAwait(false);
		}
		catch (SocketException) when (this.disposeCancellationSource.IsCancellationRequested)
		{
			// The server was disposed.
			return null;
		}
		catch (ObjectDisposedException) when (this.disposeCancellationSource.IsCancellationRequested)
		{
			// The server was disposed.
			return null;
		}

		tcpClient.Client.ConfigureSocketOptionsForSsh();

		NetworkStream stream = tcpClient.GetStream();
		return stream;
	}

	public void Dispose()
	{
		this.Dispose(true);
		GC.SuppressFinalize(this);
	}

	protected virtual void Dispose(bool disposing)
	{
		if (disposing)
		{
			this.disposeCancellationSource.Cancel();
			this.disposeCancellationSource.Dispose();

			foreach (SshServerSession serverSession in this.sessions.ToArray())
			{
				serverSession.Dispose();
			}

			lock (this.sessionsLock)
			{
				this.sessions.Clear();
			}
		}
	}
}

File: src/ts/ssh-tcp/sshServer.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as net from 'net';
import { Event, Emitter, Disposable } from 'vscode-jsonrpc';
import {
	SshServerSession,
	SshServerCredentials,
	SshSessionConfiguration,
	SshProtocolExtensionNames,
	NodeStream,
	Stream,
	SshConnectionError,
	SshDisconnectReason,
	Trace,
	TraceLevel,
	SshTraceEventIds,
} from '@microsoft/dev-tunnels-ssh';
import { TcpListenerFactory, DefaultTcpListenerFactory } from './tcpListenerFactory';

/**
 * Enables accepting SSH sessions on a TCP socket.
 *
 * It's possible to create an `SshServerSession` over any `Stream` instance;
 * this class is merely a convenient helper that manages creating sessions
 * over Node.js TCP `Socket`s from incoming connections.
 */
export class SshServer implements Disposable {
	private tcpListener?: net.Server;
	private readonly sessions: SshServerSession[] = [];
	private readonly reconnectableSessions: SshServerSession[] | undefined;

	public constructor(private config: SshSessionConfiguration) {
		if (!config) throw new TypeError('SshSessionConfiguration is required.');

		if (config.protocolExtensions.includes(SshProtocolExtensionNames.sessionReconnect)) {
			this.reconnectableSessions = [];
		}
	}

	/**
	 * Gets or sets a function that handles trace messages associated with the server sessions.
	 *
	 * By default, no messages are traced. To enable tracing, set this property to a function
	 * that routes the message to console.log, a file, or anywhere else.
	 *
	 * @param level The level of message being traced: error, warning, info, or verbose.
	 * @param eventId An integer that identifies the type of event. Normally this is one of
	 * the values from `SshTraceEventIds`, but extensions may define additional event IDs.
	 * @param msg A description of the event (non-localized).
	 * @param err Optional `Error` object associated with the event, often included with
	 * warning or error events. While the `Error.message` property is typically included as
	 * (part of) the `msg` parameter, the error object may contain additional useful context
	 * such as the stack trace.
	 */
	public trace: Trace = (level, eventId, msg, err) => {};

	private readonly errorEmitter = new Emitter<Error>();
	public readonly onError: Event<Error> = this.errorEmitter.event;

	private readonly sessionOpenedEmitter = new Emitter<SshServerSession>();
	public readonly onSessionOpened: Event<SshServerSession> = this.sessionOpenedEmitter.event;

	public readonly credentials: SshServerCredentials = { publicKeys: [] };

	/**
	 * Gets or sets a factory for creating TCP listeners.
	 *
	 * Applications may override this factory to provide custom logic for selecting
	 * local port numbers to listen on for port-forwarding.
	 */
	public tcpListenerFactory: TcpListenerFactory = new DefaultTcpListenerFactory();

	public async acceptSessions(localPort: number, localAddress?: string): Promise<void> {
		if (!localAddress) {
			localAddress = '0.0.0.0';
		}

		const portPrefix = localAddress === '0.0.0.0' ? 'port ' : localAddress + ':';

		try {
			this.tcpListener = await this.tcpListenerFactory.createTcpListener(
				undefined, // remotePort
				localAddress,
				localPort,
				false,
			);
		} catch (e) {
			if (!(e instanceof Error)) throw e;
			this.trace(
				TraceLevel.Error,
				SshTraceEventIds.serverListenFailed,
				`SshServer failed to listen on ${portPrefix}${localPort}: ${e.message}`,
				e,
			);
			throw e;
		}

		this.tcpListener.addListener('connection', this.acceptSession.bind(this));
		this.trace(
			TraceLevel.Info,
			SshTraceEventIds.serverListening,
			`SshServer listening on ${portPrefix}${localPort}.`,
		);
	}

	protected async acceptConnection(socket: net.Socket): Promise<Stream> {
		socket.setNoDelay(true);
		return new NodeStream(socket);
	}

	private async acceptSession(socket: net.Socket): Promise<void> {
		this.trace(
			TraceLevel.Info,
			SshTraceEventIds.serverClientConnected,
			'SshServer client connected.',
		);

		const stream = await this.acceptConnection(socket);
		const session = new SshServerSession(this.config, this.reconnectableSessions);
		session.trace = this.trace;
		session.credentials = this.credentials;
		this.sessions.push(session);

		session.onClosed((e) => {
			const sessionIndex = this.sessions.indexOf(session);
			if (sessionIndex >= 0) {
				this.sessions.splice(sessionIndex, 1);
			}
		});
		this.sessionOpenedEmitter.fire(session);

		try {
			await session.connect(stream);
		} catch (e) {
			if (!(e instanceof Error)) throw e;
			if (e instanceof SshConnectionError) {
				await session.close(e.reason || SshDisconnectReason.connectionLost, e.message, e);
			} else {
				await session.close(SshDisconnectReason.protocolError, e.message, e);
			}
			this.errorEmitter.fire(e);
		}
	}

	public dispose(): void {
		this.tcpListener?.close();
	}
}
